File: Sources/Buckett/Services/UploadHistory.swift
import Foundation

/// Local record of upload activity per day, powering the Statistics heatmap.
/// Purely local — nothing is sent anywhere.
@MainActor
final class UploadHistory: ObservableObject {
    static let shared = UploadHistory()

    struct Day: Codable {
        var files: Int = 0
        var bytes: Int64 = 0
    }

    /// Keyed by "yyyy-MM-dd" (local time zone).
    @Published private(set) var days: [String: Day] = [:]

    private var fileURL: URL {
        AccountStore.supportDirectory.appendingPathComponent("upload-history.json")
    }

    private static let keyFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    private init() {
        if let data = try? Data(contentsOf: fileURL),
           let decoded = try? JSONDecoder().decode([String: Day].self, from: data) {
            days = decoded
        }
    }

    static func key(for date: Date) -> String {
        keyFormatter.string(from: date)
    }

    func record(bytes: Int64, date: Date = Date()) {
        let key = Self.key(for: date)
        var day = days[key] ?? Day()
        day.files += 1
        day.bytes += bytes
        days[key] = day
        save()
    }

    func day(for date: Date) -> Day? {
        days[Self.key(for: date)]
    }

    /// The last `count` days, oldest first.
    func recentDays(_ count: Int) -> [(date: Date, day: Day?)] {
        let calendar = Calendar.current
        let today = calendar.startOfDay(for: Date())
        return (0..<count).reversed().compactMap { offset in
            guard let date = calendar.date(byAdding: .day, value: -offset, to: today) else {
                return nil
            }
            return (date, day(for: date))
        }
    }

    private func save() {
        try? FileManager.default.createDirectory(
            at: AccountStore.supportDirectory, withIntermediateDirectories: true
        )
        if let data = try? JSONEncoder().encode(days) {
            try? data.write(to: fileURL, options: .atomic)
        }
    }
}

File: windows/src/Buckett/Services/UploadHistory.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json;
using CommunityToolkit.Mvvm.ComponentModel;

namespace Buckett.Services;

/// Local record of upload activity per day, powering the Statistics heatmap.
/// Purely local — nothing is sent anywhere.
public sealed class UploadHistory : ObservableObject
{
    public static UploadHistory Shared { get; } = new();

    public sealed class Day
    {
        public int files { get; set; }
        public long bytes { get; set; }
    }

    /// Keyed by "yyyy-MM-dd" (local time zone).
    private Dictionary<string, Day> _days = new();

    public IReadOnlyDictionary<string, Day> Days => _days;

    private static string FilePath => Path.Combine(AppPaths.SupportDirectory, "upload-history.json");

    private UploadHistory()
    {
        try
        {
            if (File.Exists(FilePath))
            {
                var json = File.ReadAllText(FilePath);
                var decoded = JsonSerializer.Deserialize<Dictionary<string, Day>>(json);
                if (decoded != null) _days = decoded;
            }
        }
        catch (Exception error)
        {
            Log.Warn($"failed to load upload history: {error.Message}");
        }
    }

    public static string Key(DateTime date) =>
        date.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);

    public void Record(long bytes, DateTime? date = null)
    {
        var key = Key(date ?? DateTime.Now);
        if (!_days.TryGetValue(key, out var day))
        {
            day = new Day();
            _days[key] = day;
        }
        day.files += 1;
        day.bytes += bytes;
        Save();
        OnPropertyChanged(nameof(Days));
    }

    public Day? DayFor(DateTime date) => _days.TryGetValue(Key(date), out var day) ? day : null;

    /// The last `count` days, oldest first.
    public List<(DateTime Date, Day? Day)> RecentDays(int count)
    {
        var today = DateTime.Today;
        return Enumerable.Range(0, count)
            .Reverse()
            .Select(offset =>
            {
                var date = today.AddDays(-offset);
                return (date, DayFor(date));
            })
            .ToList();
    }

    private void Save()
    {
        try
        {
            AppPaths.EnsureSupportDirectory();
            File.WriteAllText(FilePath, JsonSerializer.Serialize(_days));
        }
        catch (Exception error)
        {
            Log.Warn($"failed to save upload history: {error.Message}");
        }
    }
}
